
---
title: "PACSI Tutorial"
author: "Chonghui Liu<br/>Guohua Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    toc: yes
vignette: |
  %\VignetteIndexEntry{PACSI Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
---


# Introduction

PACSI, a novel network-based method to identify cell subpopulations associated with disease phenotypes of interest.

```{r message=FALSE}
library(PACSI)
```

# single-cell data example

The input of PACSI are: a single-cell expression matrix, a bulk expression matrix, a ppi data and a phenotype label data. The phenotype label of each bulk sample should be binary. In this tutorial, we use a example to show how to execute PACSI in real applications.


### Prepare the scRNA-seq data 

We load in the HNSC scRNA-seq data:
```{r}
data("case1_sc_data_1to100")
sc_data <- case1_sc_data_1to100
```
In this scRNA-seq data, each row represents a gene and each column represents a cell.The single-cell expression matrix should be TPM/FPKM-normalized. The dimensions of the single-cell data are:
```{r}
dim(sc_data)
```
which indicates there are 5,000 genes and 100 cells in total. For the scRNA-seq data used in PACSI, We can use the functions in the Seurat package to preprocess this data:
```{r warning=FALSE}
library(Seurat)
sc_data2 <- Seurat::CreateSeuratObject(as.matrix(sc_data))
sc_data2 <- Seurat::FindVariableFeatures(object = sc_data2, verbose = F, nfeatures=2000)
sc_data2 <- Seurat::ScaleData(object = sc_data2, verbose = F)
sc_data2 <- Seurat::RunPCA(object = sc_data2, features = Seurat::VariableFeatures(sc_data2), verbose = TRUE)
sc_data2 <- Seurat::FindNeighbors(object = sc_data2, dims = 1:5, verbose = TRUE)
sc_data2 <- Seurat::FindClusters( object = sc_data2, resolution = 0.6, verbose = TRUE)
sc_data2 <- Seurat::RunUMAP(object = sc_data2, dims = 1:5, verbose = TRUE)
```
we can visualize these 100 cells using the UMAP coordinates:
```{r fig.align="center", fig.width = 5.5}
Seurat::DimPlot(sc_data2, reduction = 'umap', label.size = 5)
```

### Prepare the bulk data, phenotype labels and PPI data

We load in the preprocessed HNSC bulk expression matrix and the corresponding phenotype data.And then we locate the PPI data.
```{r}
data("case1_bulk_data_200to300")
bulk_data <- case1_bulk_data_200to300
data("case1_sample_pheno_labels_200to300")
phenotype_data <- case1_sample_pheno_labels_200to300
ppi_location <- system.file("extdata", "ppi.csv", package = "PACSI")
```
In this bulk data, each row represents a gene and each column represents a sample. The bulk expression matrix should be TPM/FPKM-normalized. The dimensions of the bulk data are:
```{r}
dim(bulk_data)
```
It shows this data has 5000 genes and 101 samples in total. Users do not need to keep the common genes between the single-cell and bulk samples. Besides, all of these samples have phenotype labels:
```{r}
head(phenotype_data)
```
The location of PPI data is:
```{r}
ppi_location
```

### Execute PACSI to select the cells associated with disease phenotype

Given the above inputs, we can use PACSI to select the phenotype-associated cell subpopulations:
```{r warning=FALSE}
PACSI_result <- PACSI(bulk_data, phenotype_data, sc_data, ppi_data = ppi_location,times = 10,ncores = 8)
```

```{r}
head(PACSI_result$result)
```
We can look at the results of PACSI.:
```{r warning=FALSE}
table(PACSI_result$result$cell_phenotype_labels)
```

We can visualize PACSI results, where 1 represents the phenotype-associated cells selected by PACSI and 0 represents other cells:
```{r warning=FALSE, fig.align="center", fig.width = 5.5}
sc_data3 <- AddMetaData(sc_data2, metadata = PACSI_result[["result"]]$cell_phenotype_labels, col.name="PACSI_predict")
DimPlot(sc_data3, reduction = 'umap', group.by = 'PACSI_predict', cols = c('grey','indianred1'))
```

At last, we can get access to the parameters using:
```{r}
PACSI_result[["parameter"]]
```



# Spatial transcriptomic data example
The input of PACSI are:  a spatial transcriptomic profile, a bulk expression matrix, a ppi data and a phenotype label data.  In this tutorial, we use a example to show how to execute PACSI in real applications.

### Prepare the ST data 
PACSI accepts a Seurat object of spatial transcriptomics data. The transcriptomics data needs to be pre-processed.  This means that Seurat pipelines such as normalization, dimensional reduction, clustering, and visualization should be applied prior to the analysis with PACSI. We load in the spatial transcriptomics data:
```{r}
library(Seurat)
load("BRCA_ST.rda")
```
we can visualize these data using the seurat package:
```{r warning=FALSE}
SpatialDimPlot(BRCA_ST, label = TRUE, label.size = 3)
```

### Prepare the bulk data, phenotype labels and PPI data

We load in the preprocessed bulk expression matrix and the corresponding phenotype data.And then we locate the PPI data.
```{r warning=FALSE}
library(PACSI)
data("case1_bulk_data_200to300")
bulk_data <- case1_bulk_data_200to300
data("case1_sample_pheno_labels_200to300")
phenotype_data <- case1_sample_pheno_labels_200to300
ppi_location <- system.file("extdata", "ppi.csv", package = "PACSI")
```
### Execute PACSI to select the spots associated with disease phenotype

Given the above inputs, we can use PACSI to select the phenotype-associated spos:
```{r warning=FALSE}
PACSI_ST_result <- PACSI(bulk_data, phenotype_data, BRCA_ST, ppi_data=ppi_location, spatial=T, times = 10, ncores = 8)
```
We can visualize PACSI results, where 1 represents the phenotype-associated spots selected by PACSI and 0 represents other spots::
```{r}
BRCA_ST <- AddMetaData(BRCA_ST, metadata=PACSI_ST_result[["result"]]$cell_phenotype_labels, col.name="PACSI_predict")
SpatialDimPlot(BRCA_ST, label = F,group.by = 'PACSI_predict', cols=c('grey','indianred1'))
```
